subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationResult.java

```diff
@@ -151,6 +194,50 @@ public File getOutputLocation() {
             public File resolveForInputArtifact(File inputArtifact) {
                 return outputFile;
             }
+
+            @Override
+            public void visitResult(TransformationResultVisitor visitor) {
+                visitor.visitOutput(outputFile);
+            }
+        }
+    }
+
+    interface TransformationResultVisitor {
+        void visitInput();
+        void visitInput(String relativePath);
+        void visitOutput(File outputLocation);
+    }
+
+    class OutputTypeInferringBuilder {
+        private final File inputArtifact;
+        private final File outputDir;
+        private final String inputArtifactPrefix;
+        private final String outputDirPrefix;
+        private final Builder delegate = TransformationResult.builder();
+
+        public OutputTypeInferringBuilder(File inputArtifact, File outputDir) {
+            this.inputArtifact = inputArtifact;
+            this.outputDir = outputDir;
+            this.inputArtifactPrefix = inputArtifact.getPath() + File.separator;
+            this.outputDirPrefix = outputDir.getPath() + File.separator;
+        }
+
+        public void addOutput(File output, Consumer<File> workspaceAction) {
```

Comment:
Why pass `workspaceAction` here instead of calling it in `DefaultTransformOutputs.resolveAndRegister()`?

Reply:
Because we only call the action when the file is in the workspace, and I didn't want to return the enum we had before here. I you think returning the enum is better, then I can go back to that.